Modification of graph traversal rules for node export and deletion (#3459)
aiidateam/aiida-core: docs/source/concepts/provenance.rst

Analogously, when deleting a process, the parent workflow that has called it (if present) will be deleted as well (by traversing a ``backward`` ``call_calc`` or ``call_work`` link).
Since the traversal rules are applied recursively, this means that also the caller of the caller of the process will be deleted, and so on.

The possibility to follow ``call`` links in the other direction is enabled by default, which causes that targeting a given process for either export or deletion will also include in the process not only all of his child processes, but also all the childs of any of its parent processes as well.
As a result of all ``call`` links being traversed in both directions, targeting any of the process nodes in a workflow will mean the inclusion of the other processes of that workflow as well.

> The option of disabling the traversal of the ``call`` links in one of the directions (forward if deleting, backward if exporting) provides the users with the ability to select parts within their workflows to affect, but this usually requires a much deeper understanding of the conectivity of the nodes and the complex behavior that arises from the combination of the different traversal rules (see below).

Reviewer: ```suggestion
Users can disable the traversal of ``call`` links in one of the directions (forward for deletion, backward for export) for fine-grained control (see examples below).
```

However, this usually requires a deeper understanding of the traversal rules and may imply a more thorough analysis of the particular graph.
To better illustrate this, we will now consider the aplication of the deletion procedure to the following graph:

.. _delexp_example02:
.. image:: include/images/delexp_example02.png
   :scale: 80%

As you can see, |W_1| and |W_2| describe two similar but independent procedures that were launched by a single parent workflow |W_0|.
A typical user would have obtained this by directly running this workflow |W_0| to obtain the results |D_3| and |D_4| from the inputs |D_1| and |D_2|, and may even be unaware of the internal division of |W_0| into two sub-Workflows |W_1| and |W_2|.
Hence, if the user considers the workflow (meaning, the whole set of nodes produced by it) no longer necessary, the intuitive thing to do in order to remove it from its database would be by targeting the workflow node |W_0| for deletion.
Indeed, this would produce the desired result:

.. _delexp_example02-a01:

> .. image:: include/images/delexp_example02-a01.png

Reviewer: Is there a need for this figure, where both D3 and D4 are targeted?

To me it would seem to suffice (and make things shorter) to keep only the next figure, which is identical except that only D3 is targeted.

   :scale: 80%

The nodes |W_1| and |W_2| would be included because |W_0| is being targeted (``call_work_forward=True``), then the nodes |C_1| and |C_2| would also be included (``call_calc_forward=True``), and finally the nodes |D_3| and |D_4| would end up being included as well (``create_forward=True``).
In the end, only the inputs |D_1| and |D_2| remain (since ``input_work_backward=False`` always and ``input_calc_backward=False`` by default).

> The same result would occurr if the user were to target the output nodes instead (intending to delete everything associated with the obtention of those results):

Reviewer: ```suggestion
The same result would occur if the user were to target the output nodes instead (intending to delete everything associated with the obtention of those results):
```

   :scale: 80%

In this case, the nodes |C_1| and |C_2| would first be included because the data nodes |D_3| and |D_4| are being targeted (``create_reverse=True``), and this would then include the nodes |W_1| and |W_2| (``call_calc_reverse=True``) and finally the parent workflow |W_0| (``call_work_reverse=True``).
It is important to notice that even if the user deletes only one of the outputs, the whole set of nodes generated by the workflow would be deleted, and not just the ones associated to the targeted data node.
As the results |D_3| and |D_4| where obtained from the same high-level process |W_0|, then the default behavior has the underlying assumption that they are interconected and not independent from one another (as if they were two different outputs of a single calculation).

.. _delexp_example02-a03:
.. image:: include/images/delexp_example02-a03.png
   :scale: 80%

This results from a combination of the rules mentioned in the previous two paragraphs: so the first section (|D_3| to |C_1| to |W_1| to |W_0|) is the same as the last case viewed, and then the rest (|W_0| to |W_2| to |C_2| to |D_4|) is the same as the case when the target is |W_0|.

> This dependency between nodes becomes particularly relevant when, for example, a user with more knowledge of the internal procedures of the parent workflow |W_0| wants to only delete the calculations and results associated to workflow |W_1|.

Reviewer: I suggest you introduce a new sub-section before this sentence - e.g.

```
Deleting sub-workflows
------------------------

Say, a user with knowledge of the internal workings of the parent workflow |W_0| wants to delete *only* the calculations and results associated to workflow |W_1|.
With default traversal rules, targeting |W_1| does not produce the desired outcome:
```

The intuitive action of targeting |W_1| does not produce the desired outcome:

.. _delexp_example02-b01:
.. image:: include/images/delexp_example02-b01.png
   :scale: 80%

Indeed |C_1| and |D_4| will be deleted (through ``call_calc_forward`` from |W_1| to |C_1| and ``create_forward`` from |C_1| to |D_3|), but so will |W_0| (through ``call_work_reverse`` from |W_1|), |W_2| (``call_work_forward`` from |W_0|), |C_2| (``call_calc_forward`` from |W_2|) and |D_4| (``create_forward`` from |C_2|).

> The way to achieve the desired outcome is not at all obvious, although in this particular case one could propose some simple 'hacks', like targeting |W_1| setting the toggable flag ``call_work_forward=False`` (preventing the traversal from |W_0| to |W_2|):

Reviewer: ```suggestion
In this particular case, deletion of |W_2| could be avoided by setting the flag ``call_work_forward=False`` (preventing the traversal from |W_0| to |W_2|):
```

.. _delexp_example02-b02:
.. image:: include/images/delexp_example02-b02.png
   :scale: 80%

> However, these kind of solutions are not very generalizable (for example, this would no longer work if |W_1| also had other sub-workflows that needed to be deleted as well).

Reviewer: ```suggestion
However, this approach is not generally applicable, and wouldn't work if |W_1| had sub-workflows that needed to be deleted as well.
```

> A better approach would be to first delete, using the most conservative setup (disabling all toggable rules), the node that is linking the workflows that the user wishes to now treat separatedly (in this case |W_0|, which is linking |W_1| and |W_2|).

Reviewer: ```suggestion
A more general approach is to first sever the connection to |W_2| by deleting node |W_0| with all switchable traversal rules turned off.
```

> Then, once the independence of these two workflows (|W_1| and |W_2|) is explicitly reflected in the graph, the node |W_1| can be targeted for deletion with the default settings and nothing will happen to the procedures in |W_2|.

Reviewer: ```suggestion
Then, once the independence of |W_1| and |W_2| is explicitly reflected in the graph, delete node |W_1| with the default settings.
```

.. _delexp_example02-b03:
.. image:: include/images/delexp_example02-b03.png
   :scale: 80%

> It is worth noticing that if the workflow |W_0| was itself a sub-process of a more complicated workflow, all that high-level logic will be deleted and lost due to the non-toggable rule ``call_work_reverse=True``.

Reviewer: ```suggestion
It is worth noting that if the workflow |W_0| was itself part of a higher-level workflow, all that higher-level logic would be deleted due to the non-switchable rule ``call_work_reverse=True``.
```

This is an inevitable outcome of deleting part of a workflow, since due to the loss of that information it has become incomplete and it makes no sense to keep it.

> The number of possible scenarios and desired outcomes is impossible to cover entirely, but hopefully this example helped to show how you need to analyze the outcome of applying the delete or export procedures in your own cases of interest, especially when not using the default rules.

Reviewer: ```suggestion

```

Then nodes |W_2|, |C_2| and |D_4| will be included because |W_0| was included, for the same reasons that were explained in the paragraphs above.

Customizing the section to delete (or export)
.............................................

This dependency between nodes becomes particularly relevant when, for example, a user with more knowledge of the internal procedures of the parent workflow |W_0| wants to only delete the calculations and results associated to workflow |W_1|.
The intuitive action of targeting |W_1| does not produce the desired outcome:

.. _delexp_example02-b01:
.. image:: include/images/delexp_example02-b01.png
   :scale: 80%

Indeed |C_1| and |D_4| will be deleted (through ``call_calc_forward`` from |W_1| to |C_1| and ``create_forward`` from |C_1| to |D_3|), but so will |W_0| (through ``call_work_reverse`` from |W_1|), |W_2| (``call_work_forward`` from |W_0|), |C_2| (``call_calc_forward`` from |W_2|) and |D_4| (``create_forward`` from |C_2|).

> The way to achieve the desired outcome is not at all obvious, although in this particular case one could propose some simple 'hacks', like targeting |W_1| setting the switchable flag ``call_work_forward=False`` (preventing the traversal from |W_0| to |W_2|):

Reviewer: I did have a suggestion for this sentence as well that removed the "not at all obvious" and "hacks", which I would probably still prefer...

Author: I have now changed the "hack" and "not obvious" phrasings. I used a different version of what you proposed, not out of stuborness, but because I keep missing comments/corrections when github hides/colapses them (only now I have noticed it; this is also the reason why I didn't adress it in the previous commit). Luckily I think in this case I still adressed what you wanted change, correct? I will try to pay more attention to that in the future, apologies.

Reviewer: thanks & no worries, happens to me as well!
